cmd/minishift/cmd/start_preflight.go

```diff
@@ -0,0 +1,278 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/docker/machine/libmachine/drivers"
+	startFlags "github.com/minishift/minishift/cmd/minishift/cmd/config"
+	miniutil "github.com/minishift/minishift/pkg/minishift/util"
+
+	"github.com/minishift/minishift/pkg/util/os/atexit"
+	"github.com/spf13/viper"
+)
+
+const (
+	StorageDisk = "/mnt/sda1"
+)
+
+type preflightCheckRunner func() bool
+type preflightCheckWithDriverRunner func(driver drivers.Driver) bool
+
+func preflightCheckSucceedsOrFails(configNameOverrideIfSkipped string, execute preflightCheckRunner, message string, treatAsWarning bool, configNameOverrideIfWarning string, errorMessage string) {
+	fmt.Printf("-- %s ... ", message)
+
+	isSkip := viper.GetBool(configNameOverrideIfSkipped)
+	isWarn := viper.GetBool(configNameOverrideIfWarning)
+
+	if isSkip {
+		fmt.Println("SKIP")
+		return
+	}
+
+	if execute() {
+		fmt.Println("OK")
+		return
+	}
+
+	fmt.Println("FAIL")
+	errorMessage = fmt.Sprintf("   %s", errorMessage)
+	if isWarn {
+		fmt.Println(errorMessage)
+	} else {
+		atexit.ExitWithMessage(1, errorMessage)
+	}
+}
+
+func preflightCheckSucceedsOrFailsWithDriver(configNameOverrideIfSkipped string, execute preflightCheckWithDriverRunner, driver drivers.Driver, message string, treatAsWarning bool, configNameOverrideIfWarning string, errorMessage string) {
+	fmt.Printf("-- %s ... ", message)
+
+	isSkip := viper.GetBool(configNameOverrideIfSkipped)
+	isWarn := viper.GetBool(configNameOverrideIfWarning)
+
+	if isSkip {
+		fmt.Println("SKIP")
+		return
+	}
+
+	if execute(driver) {
+		fmt.Println("OK")
+		return
+	}
+
+	fmt.Println("FAIL")
+	errorMessage = fmt.Sprintf("   %s", errorMessage)
+	if isWarn || treatAsWarning {
+		fmt.Println(errorMessage)
+	} else {
+		atexit.ExitWithMessage(1, errorMessage)
+	}
+}
+
+func preflightChecksBeforeStartHost(vmDriver string) {
+	switch vmDriver {
+	case "xhyve":
+		preflightCheckSucceedsOrFails(
+			startFlags.SkipCheckXHyveDriver.Name,
+			checkXhyveDriver,
+			"Checking if 'xhyve' driver installed",
+			false, startFlags.WarnCheckXHyveDriver.Name,
+			"See install guide")
+	case "kvm":
+		preflightCheckSucceedsOrFails(
+			startFlags.SkipCheckKVMDriver.Name,
+			checkKvmDriver,
+			"Checking if 'kvm' driver installed",
+			false, startFlags.WarnCheckXHyveDriver.Name,
+			"See install guide")
+	case "hyperv":
+		preflightCheckSucceedsOrFails(
+			startFlags.SkipCheckHyperVDriver.Name,
+			checkHypervDriver,
+			"Checking if 'hyperv' driver configured",
+			false, startFlags.WarnCheckHyperVDriver.Name,
+			"Hyper-V virtual switch not set")
+	}
+}
+
+func checkXhyveDriver() bool {
+	_, err := exec.LookPath("docker-machine-driver-xhyve")
```

Comment:
@gbraad Interesting for me it worked on Mac also, I am not sure if because of version of MacOS I am using.

```
➜  /tmp ls -l | grep testdir 
drwxr-xr-t  2 prkumar  wheel       68 Jul 27 18:09 testdir

➜  test sw_vers 
ProductName:	Mac OS X
ProductVersion:	10.12.5
BuildVersion:	16F73

➜  test go run test.go
Sticky bit set
```